lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -728,6 +728,20 @@ CheckRedeclarationRequest::evaluate(Evaluator &eval, ValueDecl *current,
       auto found = nominal->lookupDirect(current->getBaseName(), SourceLoc(),
                                          flags);
       otherDefinitions.append(found.begin(), found.end());
+
+      // Look into the generics of the type. Value generic parameters can appear
+      // as static members of the type.
+      if (auto genericDC = static_cast<Decl *>(nominal)->getAsGenericContext()) {
+        auto gpList = genericDC->getGenericParams();
+
+        if (gpList && !current->getBaseName().isSpecial()) {
+          auto gp = gpList->lookUpGenericParam(current->getBaseIdentifier());
+
+          if (gp && gp->isValue()) {
+            otherDefinitions.push_back(gp);
```

Comment:
The case where a nested type shadows an outer value generic parameter seems to be allowed by this logic, but it's untested. Can you add a test?

Reply:
This is actually diagnosed as a redeclaration error:
```
/Users/alejandro/Desktop/swift/swift/test/Sema/value_generics.swift:185:14: error: unexpected note produced: 'I' previously declared here
struct H<let I: Int> {
             ^
/Users/alejandro/Desktop/swift/swift/test/Sema/value_generics.swift:186:10: error: unexpected error produced: invalid redeclaration of 'I'
  struct I {}
         ^
```
I've gone ahead and added a test case proving it as such.